lib/actions/write.js

```diff
@@ -1,104 +1,96 @@
-let rowCount = 0;
-let ax;
-let putUrl;
+const rawData = []
+let timeout
 
-let readyFlag = false;
+async function proceedData(data, cfg) {
+  let csvString
+  const delimiter = cfg.separator ? cfg.separator : ','
 
-async function init(cfg) {
-  const delimiter = cfg.writer.separator || ',';
-  const header = cfg.includeHeaders !== 'No';
-  logger.trace('Using delimiter: \'%s\'', delimiter);
-  const options = {
-    header,
-    delimiter,
-  };
-
-  if (cfg.writer.columns) {
-    const columns = Object.keys(_.keyBy(cfg.writer.columns, 'property'));
-    options.columns = columns;
+  const unparseOptions = {
+    header: cfg.header,
+    delimiter
   }
 
-  stringifier = csv.stringify(options);
-  signedUrl = await client.resources.storage.createSignedUrl();
-  putUrl = signedUrl.put_url;
-  logger.trace('Uploading CSV file');
-  ax = axios.create();
-  util.addRetryCountInterceptorToAxios(ax);
-  readyFlag = true;
-}
-async function ProcessAction(msg, cfg) {
-  // eslint-disable-next-line consistent-this
-  const self = this;
-
-  while (!readyFlag) {
-    // eslint-disable-next-line no-loop-func,no-await-in-loop
-    await new Promise((resolve) => timeout(resolve, 100));
+  if (cfg.order) {
+    // create fields array from string
+    // eslint-disable-next-line
+    const fields = cfg.order.split(`${delimiter}`).map(x => { return x.trim() })
+    const orderedData = data.map((value) => {
+      const result = fields.map((key) => {
+        const filtered = value[key]
+        return filtered
+      })
+      return result
+    })
+    csvString = papa.unparse({
+      fields,
+      data: orderedData
+    }, unparseOptions)
+  } else {
+    csvString = papa.unparse(data, unparseOptions)
   }
 
-  if (timeout) {
-    clearTimeout(timeout);
+  if (!cfg.uploadToAttachment) {
+    await this.emit('data', messages.newMessageWithBody({ csvString }))
+    this.logger.info(`Complete, memory used: ${process.memoryUsage().heapUsed / 1024 / 1024} Mb`)
+    return
   }
 
-  timeout = setTimeout(async () => {
-    readyFlag = false;
-
-    self.logger.info('Closing the stream due to inactivity');
-
-    const finalRowCount = rowCount;
-    self.logger.info('The resulting CSV file contains %s rows', finalRowCount);
-    ax.put(putUrl, stringifier, {
-      method: 'PUT',
-      timeout: REQUEST_TIMEOUT,
-      retry: REQUEST_MAX_RETRY,
-      delay: REQUEST_RETRY_DELAY,
-      maxContentLength: REQUEST_MAX_CONTENT_LENGTH,
-    });
-    stringifier.end();
+  const attachmentProcessor = new AttachmentProcessor()
+  let attachment
+  try {
+    attachment = await attachmentProcessor.uploadAttachment(csvString)
+  } catch (err) {
+    this.logger.error(`Upload attachment failed: ${err}`)
+    this.emit('error', `Upload attachment failed: ${err}`)
+  }
+  const body = {
+    attachmentUrl: attachment.config.url,
+    type: '.csv',
+    size: Buffer.byteLength(csvString),
+    attachmentCreationTime: new Date(),
+    attachmentExpiryTime: new Date(),
```

Comment:
I would remove `attachmentExpiryTime` for right now. In the long term, we will migrate attachments to the Maester service which will have an expiry time. Howver, without that migration, the field `attachmentExpiryTime` doesn't make sense.

Reply:
Done